server/service/mdm.go

```diff
@@ -2444,3 +2445,259 @@ func (svc *Service) DeleteMDMAppleAPNSCert(ctx context.Context) error {
+	if err != nil {
+		return "", false, fmt.Errorf("creating request to Apple VPP endpoint: %w", err)
+	}
+
+	req.Header.Add("Authorization", bearer)
+
+	client := fleethttp.NewClient(fleethttp.WithTimeout(10 * time.Second))
+	resp, err := client.Do(req)
+	if err != nil {
+		return "", false, fmt.Errorf("making request to Apple VPP endpoint: %w", err)
+	}
+	defer resp.Body.Close()
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return "", false, fmt.Errorf("reading response body from Apple VPP endpoint: %w", err)
+	}
+
+	// For some reason, Apple returns 200 OK even if you pass an invalid token in the Auth header.
+	// We will need to parse the response and check to see if it contains an error.
+
+	var respJSON struct {
+		LocationName string `json:"locationName"`
+		ErrorNumber  int    `json:"errorNumber"`
+	}
+
+	if err := json.Unmarshal(body, &respJSON); err != nil {
+		return "", false, fmt.Errorf("parsing response body from Apple VPP endpoint: %w", err)
+	}
+
+	// Per https://developer.apple.com/documentation/devicemanagement/app_and_book_management/app_and_book_management_legacy/interpreting_error_codes
+	if resp.StatusCode == 401 || respJSON.ErrorNumber == 9622 {
+		return "", false, nil
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return "", false, fmt.Errorf("calling Apple VPP config endpoint failed with status %d", resp.StatusCode)
+	}
+
+	return respJSON.LocationName, true, nil
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// GET /vpp
+////////////////////////////////////////////////////////////////////////////////
+
+type getMDMAppleVPPTokenRequest struct{}
+
+type getMDMAppleVPPTokenResponse struct {
+	*fleet.VPPTokenInfo
+	Err error `json:"error,omitempty"`
+}
+
+func (r getMDMAppleVPPTokenResponse) error() error {
+	return r.Err
+}
+
+func getMDMAppleVPPTokenEndpoint(ctx context.Context, request interface{}, svc fleet.Service) (errorer, error) {
+	vpp, err := svc.GetMDMAppleVPPToken(ctx)
+	if err != nil {
+		return &getMDMAppleVPPTokenResponse{Err: err}, nil
+	}
+
+	return &getMDMAppleVPPTokenResponse{VPPTokenInfo: vpp}, nil
+}
+
+func (svc *Service) GetMDMAppleVPPToken(ctx context.Context) (*fleet.VPPTokenInfo, error) {
+	if err := svc.authz.Authorize(ctx, &fleet.AppleCSR{}, fleet.ActionRead); err != nil {
+		return nil, err
+	}
+
+	r, err := svc.ds.GetAllMDMConfigAssetsByName(ctx, []fleet.MDMAssetName{fleet.MDMAssetVPPToken})
+	if err != nil {
+		return nil, ctxerr.Wrap(ctx, err, "get mdm config assets by name VPP token")
+	}
+
+	decodedBytes, err := base64.StdEncoding.DecodeString(string(r[fleet.MDMAssetVPPToken].Value))
+	if err != nil {
+		return nil, ctxerr.Wrap(ctx, err, "decoding VPP token data")
+	}
+
+	var tokenData fleet.VPPTokenData
+	if err := json.Unmarshal(decodedBytes, &tokenData); err != nil {
+		return nil, ctxerr.Wrap(ctx, err, "unmarshaling VPP token data")
+	}
+
+	decodedTokenBytes, err := base64.StdEncoding.DecodeString(tokenData.Token)
```

Comment:
are we sure this value is base64 encoded? The whole `VPPTokenData` struct is encoded as base64, but I don't think this nested `Token` value is encoded too.

I think this will show when you add the other CRUD tests, but I might be wrong (I haven't tried it)

Reply:
Hey yeah that field is base64 encoded as well, but I will change it to not be b64 encoded. So neither the outer nor the inner object will be b6 encoded; we will store everything raw (encrypted ofc).

Reply:
just for my own learning, where is it being encoded?

Reply:
It's uploaded as a b64 encoded string.

Reply:
but in `server/service/mdm.go` line `2537` we decode it, right?